benchmark/single-source/KeyPathPerformanceTests.swift

```diff
@@ -0,0 +1,1009 @@
+    tags: [.validation, .api],
+    setUpFunction: setupSingleton
+  ),
+]
+
+// Number of iterations for each benchmark.
+let numberOfElementsForNestedStructs = 20000
+
+// Make it easy to switch between primitives of different types (Int, Float, Double, etc.).
+typealias ElementType = Double
+
+// Variables used across the benchmarks
+let expectedIntForNestedStructs = 3
+let arrayHolder = FixedSizeArrayHolder()
+
+let appendedKeyPath = singleHopKeyPath0.appending(path: singleHopKeyPath1)
+  .appending(path: singleHopKeyPath2).appending(path: singleHopKeyPath3)
+  .appending(path: singleHopKeyPath4)
+
+// The purpose of this class is to hold arrays used during the course of the benchmarks.
+// This is to avoid storing them in the global scope, which can slow things down.
+// We force instantiation in the setUpFunction so that we exclude the setup time from the benchmark time itself,
+// otherwise the instance would be lazily instantiated, and thereby add to the benchmark time.
+class FixedSizeArrayHolder {
+  var fixedSizeArray100: FixedSizeArray100<ElementType>
+  var mainArrayForNestedStructs: [A]
+  static let shared = FixedSizeArrayHolder()
+  init() {
+    fixedSizeArray100 = initializeFixedSizeArray100()
+    mainArrayForNestedStructs = [A]()
+  }
+
+  func reset() {
+    fixedSizeArray100 = initializeFixedSizeArray100()
+  }
+
+  func forceInstantiation() {}
+}
+
+// Setup Functions.
+public func setupKeyPathNestedStructs() {
+  FixedSizeArrayHolder.shared.forceInstantiation()
+  for _ in 0 ..< numberOfElementsForNestedStructs {
+    let instance = A(a: 0, b: B(b: 0, c: C(c: 0, d: D(d: 0, e: E(e: expectedIntForNestedStructs)))))
+    FixedSizeArrayHolder.shared.mainArrayForNestedStructs.append(instance)
+  }
+}
+
+public func setupSingleton() {
+  FixedSizeArrayHolder.shared.forceInstantiation()
+}
+
+func computeSumOfFixedSizeArray100(fixedSizeArray100: inout FixedSizeArray100<ElementType>)
+  -> ElementType
+{
+  var sum = ElementType(0)
+  withUnsafeBytes(of: fixedSizeArray100) {
+    var pointer = $0.baseAddress.unsafelyUnwrapped
+    for _ in 0 ..< fixedSizeArray100.count {
+      sum += pointer.assumingMemoryBound(to: ElementType.self).pointee
+      pointer = pointer.advanced(by: MemoryLayout<ElementType>.size)
+    }
+  }
+  return sum
+}
+
+// Used for run_KeyPathNestedStructs().
+struct A {
+  var a: Int
+  var b: B
+}
+
+struct B {
+  var b: Int
+  var c: C
+}
+
+struct C {
+  var c: Int
+  var d: D
+}
+
+struct D {
+  var d: Int
+  var e: E
+}
+
+struct E {
+  var e: Int
+}
+
+let singleHopKeyPath0: WritableKeyPath<A, B> = \A.b
+let singleHopKeyPath1: WritableKeyPath<B, C> = \B.c
+let singleHopKeyPath2: WritableKeyPath<C, D> = \C.d
+let singleHopKeyPath3: WritableKeyPath<D, E> = \D.e
+let singleHopKeyPath4: WritableKeyPath<E, Int> = \E.e
+
+// Used for run_KeyPathFixedSizeArrayAccess() and run_testDirectAccess().
+struct FixedSizeArray100<Element>: Sequence, IteratorProtocol {
```

Comment:
What's the purpose of 100 elements if only a small subset is used in the benchmarks?

Reply:
Hello, thanks for your feedback! We're wanting to introduce a suite of benchmarks that highlight some of the performance issues we've been experience with KeyPaths. What we've seen so far can be largely placed into three categories. 

The first category involves a collection of KeyPaths not being inlined if they're spread too far apart in memory. The magic number, after some experimentation, happened to be 14. As soon as you have a 15th KeyPath, the inlining in release mode does not occur, and performance falls off a cliff -- being some 100 times worse or more. This is part of the reason why we chose a struct with 100 elements instead of just 10. (I do agree that asking the optimizer to inline KeyPaths pointing to an arbitrarily-wide range is wishful thinking, but perhaps the value of 14 is smaller than some might expect.) I've added a comment right above `run_testKeyPathsInlined()` which explains this in a bit more detail.

The second category involves nested structs or classes. Any sort of KeyPath projection operation is O(N) in the depth of the object. The benchmark `KeyPathNestedStructs` uses this.

The third category involves accessing a constant-sized value within structs of various sizes. Namely, performance decreases as more and more variables are added to the struct. I've added a benchmark called `KeyPathsSmallStruct` to showcase the difference in performance between a struct with 10 elements and one with 100 `(KeyPathReadPerformance)`.

Reply:
I'm not sure what you mean with "inlining" here. Maybe you can use `@inline(never)` to ensure you are testing the non-inlined case?

Reply:
I'd like to keep it if possible: I've added the `@inline(never)` directive to both versions of`getKeypathToElement()` -- for the 10-element struct and also for the 100-element struct -- and compared the times for each iteration. In both Debug and Release mode (i.e., after compiling via `swift build` and `swift build --configuration release`), the 100- element struct needs about 1.5X more time per iteration when compared with the 10-element struct.